### DISCUSSION TITLE
Exercise: Identifying Aggregates and Aggregate Roots in a Car Rental System

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/iSAQB-CPSA-A-DDD#2] [Hands-on]
**Exercise: Identifying Aggregates and Aggregate Roots in a Car Rental System**

---

**Scenario:**

You are developing a car rental system that manages various aspects of car rentals, including customers, reservations, vehicles, locations, and payments. The system must handle operations such as making reservations, processing rentals, handling vehicle maintenance, and managing customer memberships.

---

**Domain Models:**

1. **Customer**
   - **Attributes:** CustomerID, Name, DriverLicenseNumber, MembershipStatus
   - **Behaviors:** Update personal information, Upgrade membership

2. **Reservation**
   - **Attributes:** ReservationID, CustomerID, VehicleType, PickupLocation, DropoffLocation, PickupDate, DropoffDate, Status
   - **Behaviors:** Create reservation, Modify reservation, Cancel reservation

3. **Vehicle**
   - **Attributes:** VehicleID, Make, Model, Year, LicensePlateNumber, CurrentLocation, Status (Available, Rented, UnderMaintenance)
   - **Behaviors:** Update status, Assign to rental

4. **Rental**
   - **Attributes:** RentalID, ReservationID, CustomerID, VehicleID, PickupDate, DropoffDate, TotalCost
   - **Behaviors:** Start rental, End rental, Calculate charges

5. **Location**
   - **Attributes:** LocationID, Address, OperatingHours, VehicleInventory
   - **Behaviors:** Update location information, Manage vehicle inventory

6. **Payment**
   - **Attributes:** PaymentID, RentalID, Amount, PaymentMethod, PaymentDate
   - **Behaviors:** Process payment, Issue refund

7. **MaintenanceRecord**
   - **Attributes:** RecordID, VehicleID, MaintenanceDate, Description, Cost
   - **Behaviors:** Schedule maintenance, Complete maintenance

8. **Invoice**
   - **Attributes:** InvoiceID, RentalID, CustomerID, AmountDue, DueDate, Status
   - **Behaviors:** Generate invoice, Send invoice to customer, Update payment status

---

**Details and Hints:**

- A **Customer** can have multiple **Reservations** and **Rentals**, but each **Reservation** and **Rental** is linked to a single **Customer**.
- A **Reservation** holds a spot for a vehicle type but is assigned an actual **Vehicle** when the rental starts.
- The **Rental** is created when the customer picks up the vehicle, transitioning from a **Reservation**.
- **Payments** are associated with **Rentals**; a single rental may involve multiple payments (e.g., deposits, final charges).
- **MaintenanceRecords** are linked to **Vehicles** and track maintenance activities that affect vehicle availability.
- **Invoices** are generated per **Rental** and sent to the **Customer** after the rental period ends.
- The **Vehicle** status changes based on its availability, rental status, and maintenance needs.
- Each **Location** manages its own fleet of **Vehicles** and handles pickups and drop-offs.

---

**Business Use-Cases:**

1. **Make a Reservation:**
   - A customer searches for available vehicle types at a specific location and makes a reservation for future rental.

2. **Pick Up Vehicle (Start Rental):**
   - The customer arrives at the pickup location; the system assigns an available vehicle to the reservation, creating a rental.

3. **Return Vehicle (End Rental):**
   - The customer returns the vehicle to the drop-off location; the system finalizes the rental and generates an invoice.

4. **Process Payment:**
   - The customer makes payments related to the rental, which may include deposits, rental fees, and additional charges.

5. **Vehicle Maintenance Management:**
   - A vehicle is scheduled for maintenance; it's marked as under maintenance and removed from available inventory until ready.

6. **Upgrade Customer Membership:**
   - The customer upgrades their membership status, unlocking benefits like discounts or priority reservations.

---

**Task for Students:**

Using the provided information:

1. **Identify Aggregates:**
   - Group related domain models into aggregates based on their relationships and the business invariants that need to be maintained consistently.

2. **Determine Aggregate Roots:**
   - For each aggregate, identify the root entity that will serve as the entry point for interactions with that aggregate.

3. **Justify Your Decisions:**
   - Explain the reasoning behind your grouping, considering factors such as transactional consistency, invariants, and the boundaries of each aggregate.

---

**Considerations and Hints:**

- **Transactional Consistency:** Think about which operations need to be atomic. Changes within an aggregate should be consistent and committed together.
- **Invariants:** Identify business rules that must always be consistent within an aggregate.
- **References Between Aggregates:** Aggregates can reference each other by identity but should avoid holding direct object references to maintain boundaries.
- **Size of Aggregates:** Keep aggregates as small as possible while maintaining necessary consistency.

---

**Example Starting Point:**

- **Aggregate Example:** The **Rental** process might form an aggregate that includes **Rental**, **Reservation**, **Vehicle**, and **Payment** entities, with **Rental** as the aggregate root.
- **Questions to Ponder:**
  - Should **Vehicle** be part of the **Rental** aggregate, or should it be its own aggregate?
  - How do **MaintenanceRecords** affect the **Vehicle** aggregate?
  - Is **Customer** an aggregate root, and what entities belong within its aggregate?

---

By completing this exercise, you will deepen your understanding of how to model aggregates and identify aggregate roots within a domain-driven design context, ensuring the system maintains consistency and adheres to business rules.

**Comment by akash-coded** [comment]
## Solution

---

**Solution: Identifying Aggregates and Aggregate Roots in the Car Rental System**

---

### **Identified Aggregates and Aggregate Roots:**

1. **Customer Aggregate**
   - **Aggregate Root:** **Customer**
   - **Entities within Aggregate:**
     - Customer
   - **Justification:**
     - The **Customer** entity represents individual customers with their personal information and membership status.
     - **Customer** operates independently of other entities; while it references **Reservations** and **Rentals**, they are not part of the **Customer** aggregate due to their distinct lifecycles and invariants.
     - Keeping the **Customer** aggregate focused ensures efficient updates to customer data without affecting other aggregates.

2. **Reservation Aggregate**
   - **Aggregate Root:** **Reservation**
   - **Entities within Aggregate:**
     - Reservation
   - **Justification:**
     - **Reservation** encapsulates the booking details before a rental begins.
     - It maintains its own consistency, handling creation, modification, and cancellation.
     - References to **Customer** and **VehicleType** are by IDs, maintaining aggregate boundaries.

3. **Rental Aggregate**
   - **Aggregate Root:** **Rental**
   - **Entities within Aggregate:**
     - Rental
     - Payment(s)
     - Invoice
   - **Justification:**
     - **Rental** represents the active rental period, starting when the vehicle is picked up.
     - **Payments** and **Invoices** are tightly coupled with the **Rental**, requiring transactional consistency.
     - Grouping these entities ensures all financial transactions and rental status updates occur atomically.

4. **Vehicle Aggregate**
   - **Aggregate Root:** **Vehicle**
   - **Entities within Aggregate:**
     - Vehicle
     - MaintenanceRecord(s)
   - **Justification:**
     - **Vehicle** is central to the availability and assignment process.
     - **MaintenanceRecords** directly affect the vehicle's status and availability.
     - Including **MaintenanceRecords** ensures that any changes to vehicle status consider maintenance activities, preserving invariants.

5. **Location Aggregate**
   - **Aggregate Root:** **Location**
   - **Entities within Aggregate:**
     - Location
   - **Justification:**
     - **Location** manages operational details and vehicle inventory (as IDs).
     - Vehicles themselves are not included to prevent overly large aggregates.
     - This aggregate handles updates to location information and inventory management tasks.

---

### **Aggregate Relationships and Interactions:**

- **Customer** references **Reservations** and **Rentals** by their IDs.
- **Reservation** references **CustomerID** and specifies **VehicleType**, not a specific vehicle.
- **Rental** references **ReservationID**, **CustomerID**, and **VehicleID**.
- **Vehicle** contains **MaintenanceRecords** within its aggregate.
- **Location** manages a list of **VehicleIDs** representing its inventory.

---

### **Aggregate Relationship Diagram (Textual Representation):**

```
[Customer Aggregate]
  - Customer (Aggregate Root)

[Reservation Aggregate]
  - Reservation (Aggregate Root)

[Rental Aggregate]
  - Rental (Aggregate Root)
    - Payment(s)
    - Invoice

[Vehicle Aggregate]
  - Vehicle (Aggregate Root)
    - MaintenanceRecord(s)

[Location Aggregate]
  - Location (Aggregate Root)

------------------------------------------------
Relationships (References by IDs):

Customer
  ↕
Reservation
  ↕
Rental

Rental
  ↔ (references) CustomerID
  ↔ (references) ReservationID
  ↔ (references) VehicleID

Vehicle
  ↔ (contains) MaintenanceRecord(s)

Location
  ↔ (manages) VehicleInventory [List of VehicleIDs]

Additional References:
- Reservation references VehicleType (not a specific VehicleID)
- Rental's PickupLocation and DropoffLocation reference LocationID
```

---

### **Detailed Justification of Aggregate Groupings:**

#### **1. Customer Aggregate**

- **Transactional Consistency:**
  - Operations like updating personal information and upgrading membership are self-contained.
- **Boundaries:**
  - Does not include **Reservations** or **Rentals** to prevent large aggregates and maintain clear boundaries.
- **References:**
  - Holds references to **Reservations** and **Rentals** by IDs for association purposes.

#### **2. Reservation Aggregate**

- **Transactional Consistency:**
  - Ensures that reservation details remain consistent during creation, modification, or cancellation.
- **Boundaries:**
  - Excludes **Customer** and **Vehicle** entities; they are referenced by IDs.
- **Invariants:**
  - Must maintain consistency of reservation status and associated data.

#### **3. Rental Aggregate**

- **Transactional Consistency:**
  - Starting and ending a rental, along with processing payments and generating invoices, need to be atomic.
- **Entities Included:**
  - **Payments** and **Invoice** are included to ensure financial transactions are consistent with rental operations.
- **Boundaries:**
  - **Vehicle** is referenced by ID, not included, to prevent locking vehicle data during rental updates.

#### **4. Vehicle Aggregate**

- **Transactional Consistency:**
  - Updates to vehicle status and maintenance records must be consistent.
- **Entities Included:**
  - **MaintenanceRecords** are included to directly impact vehicle availability and status.
- **Invariants:**
  - Vehicle cannot be available if under maintenance; this rule is enforced within the aggregate.

#### **5. Location Aggregate**

- **Transactional Consistency:**
  - Updates to location details and inventory management are self-contained.
- **Boundaries:**
  - Does not include **Vehicle** entities; manages inventory through a list of **VehicleIDs**.
- **Responsibilities:**
  - Handles pickups and drop-offs, ensuring operational data remains consistent.

---

### **Considerations in Aggregate Design:**

- **Size of Aggregates:**
  - Aggregates are kept small to improve performance and reduce contention.
- **Reference Handling:**
  - Entities reference other aggregates by IDs, not by holding direct object references.
- **Business Invariants:**
  - Rules that must always be true (e.g., a vehicle under maintenance cannot be rented) are enforced within the appropriate aggregate.
- **Transactional Boundaries:**
  - Operations that require atomicity are kept within the same aggregate.

---

### **Example Scenarios Demonstrating Aggregate Use:**

#### **Use-Case 1: Make a Reservation**

- **Process:**
  - Customer searches for vehicle types and creates a **Reservation**.
- **Aggregates Involved:**
  - **Reservation Aggregate**: Handles creation and consistency of the reservation.
- **Interactions:**
  - **Reservation** references **CustomerID** and desired **VehicleType**.

#### **Use-Case 2: Pick Up Vehicle (Start Rental)**

- **Process:**
  - Customer arrives; system assigns an available **Vehicle** to the **Reservation**, creating a **Rental**.
- **Aggregates Involved:**
  - **Rental Aggregate**: Creates **Rental**, processes initial payments.
  - **Vehicle Aggregate**: Updates **Vehicle** status to "Rented".
- **Interactions:**
  - **Rental** references **ReservationID**, **CustomerID**, **VehicleID**.

#### **Use-Case 3: Return Vehicle (End Rental)**

- **Process:**
  - Customer returns vehicle; system finalizes **Rental**, generates **Invoice**.
- **Aggregates Involved:**
  - **Rental Aggregate**: Updates **Rental** status, generates **Invoice**.
  - **Vehicle Aggregate**: Updates **Vehicle** status to "Available" or schedules maintenance.
- **Interactions:**
  - **Rental** processes final charges within its aggregate.
  - **Vehicle** checks for maintenance needs within its aggregate.

---

### **Conclusion:**

By structuring the system into these aggregates with clearly defined roots and boundaries, we achieve:

- **Consistency:** Business rules and invariants are enforced within aggregates.
- **Scalability:** Smaller aggregates reduce contention and improve system performance.
- **Maintainability:** Clear boundaries make the system easier to understand and modify.
- **Flexibility:** Aggregates can evolve independently as business requirements change.

---

**Comment by shruti-kirti-vwits** [comment]
https://github.com/ShrutiK-007/tinkerlab/blob/main/Aggregate.drawio
